src/contract/contract.cpp

```diff
@@ -211,3 +212,170 @@ std::string executeRain(std::string sRecipients)
+    std::string action;
+    std::string signature;
+    std::string pubkey;
+};
+
+typedef std::string MessageType_t;
+typedef std::string MessageKey_t;
+typedef std::pair<MessageType_t, MessageKey_t> MessagePair_t;
+typedef std::list<uint256> TxHashCollection;
+std::map<MessagePair_t, TxHashCollection> messageTable;
+
+bool ApplyMessage(const Message& msg, int64_t nTime)
+{
+    if (msg.action == "A")
+    {
+        /* With this we allow verifying blocks with stupid beacon */
+        if(msg.type == "beacon")
+        {
+            std::string out_cpid;
+            std::string out_address;
+            std::string out_publickey;
+            GetBeaconElements(msg.value, out_cpid, out_address, out_publickey);
+            WriteCache("beaconalt", msg.key + "." + ToString(nTime), out_publickey, nTime);
+        }
+
+        WriteCache(msg.type, msg.key, msg.value, nTime);
+        if(fDebug10 && msg.type== "beacon" )
+            LogPrintf("BEACON add %s %s %s", msg.key, DecodeBase64(msg.value), TimestampToHRDate(nTime));
+
+        if (msg.type == "poll")
+        {
+            if (Contains(msg.key, "[Foundation"))
+            {
+                msPoll = "Foundation Poll: " + msg.key.substr(0,80);
+            }
+            else
+            {
+                msPoll = "Poll: " + msg.key.substr(0,80);
+            }
+        }
+    }
+    else if(msg.action == "D")
+    {
+        if (fDebug10) LogPrintf("Deleting key type %s Key %s Value %s", msg.type, msg.key, msg.value);
+        if(fDebug10 && msg.type == "beacon" ){
+            LogPrintf("BEACON DEL %s - %s", msg.key, TimestampToHRDate(nTime));
+        }
+
+        DeleteCache(msg.type, msg.key);
+    }
+    // If this is a boinc project, load the projects into the coin:
+    else if (msg.type == "project" || msg.type == "projectmapping")
+    {
+        //Reserved
+    }
+    else
+        return false;
+
+    LogPrintf("ApplyMessage: %s %s", msg.type, msg.key);
+
+    return true;
+}
+
+bool MemorizeMessage(const CTransaction &tx)
+{
+    Message msg(tx.hashBoinc);
+
+    bool ret = ApplyMessage(msg, tx.nTime);
+    if(ret)
+    {
+        if(fDebug)
+            WriteCache("TrxID;" + msg.type, msg.key, tx.GetHash().GetHex(), tx.nTime);
+
+        messageTable[std::make_pair(msg.type, msg.key)].push_back(tx.GetHash());
+    }
+
+    return ret;
+}
+
+bool ForgetMessage(const CTransaction& tx)
+{
+    Message msg(tx.hashBoinc);
+    if(!msg.IsValid())
+        return false;
+
+    // Locate message section.
+    const uint256& tx_hash = tx.GetHash();
+    auto section_it = messageTable.find(std::make_pair(msg.type, msg.key));
+    if(section_it == messageTable.end())
+        return false;
+
+    // Locate message
+    TxHashCollection& hashes = section_it->second;
+    auto message_it = std::find(hashes.begin(), hashes.end(), tx_hash);
+    if(message_it == hashes.end())
+        return error("ForgetMessage: Hash not memorized, %s", tx_hash.GetHex());
+
+    // Invert message
```

Comment:
This construct cannot be used. We would need more intelligence in the inversion.